### PR TITLE
Add GEOSOrientPolygons

### PR DESCRIPTION
**Pull request by dbaston**
Fixes https://github.com/libgeos/geos/issues/779

### CODE REVIEW

File: capi/geos_c.h.in [hunk 1]
```diff
@@ -2723,6 +2729,18 @@ extern void GEOS_DLL GEOSGeom_setUserData(GEOSGeometry* g, void* userData);
 */
 extern int GEOS_DLL GEOSNormalize(GEOSGeometry* g);
 
+/**
+* Enforce a ring orientation on all polygonal elements in the input geometry.
+* Non-polygonal geometries will not be modified.
+*
+* \param g Input geometry
+* \param exteriorCW if 1, exterior rings will be clockwise and interior ringts
```
Comment (jorisvandenbossche) [inline capi/geos_c.h.in:2737]:
```suggestion
* \param exteriorCW if 1, exterior rings will be clockwise and interior rings
```

**Comment by jorisvandenbossche** [comment]
This is changing the input geometry (if polygonal) "inplace" as opposed to returning a new geometry? (I suppose that is consistent with existing functions like `GEOSNormalize`) 
In shapely we make our bindings to return new geometries (our general principle for user-facing API is that geometries are immutable, except for preparation), but we can easily clone on our side first as we currently do for GEOSNormalize.

**Comment by dbaston** [comment]
> This is changing the input geometry (if polygonal) "inplace" as opposed to returning a new geometry? 

I think it's a reasonable way to go for a function like this that can trivially re-use its inputs, and can be called as `GEOSOrientPolygons(GEOSClone(geom), 1)` if a copy is desired.

I guess the alternative is to have a `GEOSOrientPolygons` and `GEOSOrientPolygonsInPlace` as I think the PostGIS internals now do.

**Comment by jorisvandenbossche** [comment]
I think it is certainly fine for the C API to do it inplace (given that is how it is done for other functions, and indeed allows the user of the API to do it inplace if that is what they want, or do the clone themselves). I am not sure that warrants two different functions. 

I would maybe just call it out in the doc comment in the C API that it is changing the geometry in-place (it is somewhat implied by "Non-polygonal geometries will not be modified" (and the signature lacking a return), but could be more explicit)

**Comment by dr-jts** [comment]
Yes, mutating function should be VERY clearly identified as such in the docs.

There actually aren't very many of them, and the penalty for misuse is pretty severe.  Generally the philosophy in JTS and GEOS has been to avoid functions with side-effects.

**Comment by dr-jts** [comment]
> It's identified/enforced by the type system -- accepting a `GEOSGeometry*` rather than a `const GEOSGeometry*`.

Gotcha.

Essentially the only mutating function that exists now is `GEOSNormalize`.  In JTS this has caused enough problems that there is a non-mutating version now in place.  I think it would be better to follow the non-mutating pattern of all the other functions.

**Comment by jorisvandenbossche** [comment]
> Essentially the only mutating function that exists now is `GEOSNormalize`.

And `GEOSSetSRID` and `GEOSGeom_setUserData`

**Comment by dbaston** [comment]
I can see that in Java, where there is no language-level concept of `const`. It doesn't make sense to me in C++.

**Comment by dr-jts** [comment]
> > Essentially the only mutating function that exists now is `GEOSNormalize`.
> 
> And `GEOSSetSRID` and `GEOSGeom_setUserData`

Yes, I was putting those are in a somewhat different category - they are more like creator functions.

**Comment by pramsey** [comment]
👍 from me, merge away

**Comment by pramsey** [comment]
merge, merge, merge

**Pull request merged**